✨ KubeStellar Contributor Development Experience

dumb0002 opened the pull request:
This PR implements the KCP-Edge Contributor Development Environment

## Summary

## Related issue(s)

Additional features: https://github.com/kcp-dev/edge-mc/issues/289
Fixes # https://github.com/kcp-dev/edge-mc/issues/169

dumb0002 commented:
cc: @MikeSpreitzer @clubanderson

clubanderson commented:
/lgtm

clubanderson commented:
/lgtm

clubanderson commented:
/lgtm

MikeSpreitzer commented:
Isn't there a typo in "environments/dev-env/docs/wls_ubuntu_platform.md" --- do you want "wsl" rather than "wls"?

MikeSpreitzer commented on environments/dev-env/CONTRIBUTING.md:
```diff
@@ -0,0 +1,510 @@
+   - jq
+   - make
+   - go (version expected 1.19)
+   - kind
+   - kubectl  
```
Shouldn't we say something about version here, for consistency with the fact that kcp is based on Kubernetes version 1.24.3?

I think that exact match on version is not required; ideally we would document here the range of usable versions for each.

francostellari replied:
@MikeSpreitzer +/- 0.1 in the version is allowed. I had to force install kubectl v1.25.3 and not latest

MikeSpreitzer replied:
What sort of force are you talking about?  Installing kubectl is its own step, I am not aware of any force required to get whatever version you choose.

MikeSpreitzer commented:
I am not convinced that we want two different contributor docs, `environments/dev-env/CONTRIBUTING.md` and `CONTRIBUTING.md`.  Wouldn't it be better to have just the one, at the regular place?

MikeSpreitzer commented on environments/dev-env/CONTRIBUTING.md:
```diff
@@ -0,0 +1,510 @@
+
+
+Run the following script to install the required package (Linux or MacOS ):
+
```
I would explicitly mention the `git clone` of this repo before any usage of the local copy of this repo.

francostellari replied:
I agree

dumb0002 replied:
fixed

MikeSpreitzer commented on environments/dev-env/CONTRIBUTING.md:
````diff
@@ -0,0 +1,510 @@
+### 1. Clone the kcp-edge repo:
+
+```shell
+git clone https://github.com/kcp-dev/edge-mc.git  kcp-edge
````
(1) This has to come before the local usage of `install_req.sh` above.

(2) Must we mandate the renaming of the directory?  I think not.  This just adds extra dissonance.  Developers know that the git repo is named "edge-mc".  You might wish that the repo were named something different, and in the future it might even have a kcp-free name that we have not settled on yet, but at present the least dissonant thing to do is `git clone` without renaming.

dumb0002 replied:
(1) added a comment stating that the script needs to be run after cloning the repo. I am avoiding to add a step to run the `install_req.sh` script in our instructions -  most of the developers will already have those packages installed in their environment or they will know how to install themselves. Our `install_req.sh` is just an option that we are providing to install the required packages

(2) Fixed

MikeSpreitzer commented on environments/dev-env/CONTRIBUTING.md:
````diff
@@ -0,0 +1,510 @@
+make build
+```
+
+The kcp-edge binaries is created inside the `/bin` directory. Add this directory to your $PATH:
````
`/bin` is an absolute pathname, which is not what you mean.  Better to just say `bin`.  Or `./bin` if you really want to emphasize the locality.

MikeSpreitzer commented on environments/dev-env/CONTRIBUTING.md:
````diff
@@ -0,0 +1,510 @@
+```
+
+
+### 3. Create your own edge infrastructure (edge clusters) - Bring Your Own Cluster (BYOC)
````
This header mixes plural and singular; let's not.
"BYOC" is our own internal jargon.  I do not think we need to highlight this piece of jargon.  It is the normal usage pattern.  We should just show users doing what we expect users to do.

dumb0002 replied:
In our `demo`  environment using the `kcp-playground`, we are providing a `kind` for users with a sample workload deployed, the term BYOC clearly makes this distinction here.

MikeSpreitzer commented on environments/dev-env/CONTRIBUTING.md:
```diff
@@ -0,0 +1,510 @@
+
+### 3. Create your own edge infrastructure (edge clusters) - Bring Your Own Cluster (BYOC)
+
+Create your edge cluster or bring your own k8s edge cluster. In this example, we use [kind](https://kind.sigs.k8s.io/) to create an edge cluster that we name “florin”:
```
The heading of this section starts --- quite appropriately, IMO --- with a plural clusters.  I think the content should too.  Or at least acknowledge that the general pattern is multiple clusters and say something about starting out with just one.

dumb0002 replied:
fixed

MikeSpreitzer commented on environments/dev-env/CONTRIBUTING.md:
````diff
@@ -0,0 +1,510 @@
+kind create cluster --name florin
+```  
+
+### 4. Deploy the kcp-edge platform:
````
Steps a and b are kcp rather than kcp-edge (edge-mc).  See my earlier discussion with @francostellari about keeping clear on the fact that edge-mc is embedded in a context --- currently kcp but something more general in the future --- that provides a runtime-variable fleet of logical clusters.  That larger context is not part of edge-mc.

dumb0002 replied:
fixed

MikeSpreitzer commented on environments/dev-env/CONTRIBUTING.md:
```diff
@@ -0,0 +1,510 @@
+
+### 4. Deploy the kcp-edge platform:
+
+  a) Download kcp binaries for your platform:
```
We should probably mention the alternative of `git clone` the kcp repo and locally build and install.  This is particularly relevant for contributors.  At least for myself, I find it useful to have a local copy of the kcp code.

dumb0002 replied:
Fixed

MikeSpreitzer commented on environments/dev-env/CONTRIBUTING.md:
````diff
@@ -0,0 +1,510 @@
+
+  a) Download kcp binaries for your platform:
+
+  ```shell
````
https://github.com/kcp-dev/edge-mc/pull/344 , now merged, contributes a script for this; maybe just say to use the script?

dumb0002 replied:
fixed

MikeSpreitzer commented on environments/dev-env/CONTRIBUTING.md:
````diff
@@ -0,0 +1,510 @@
+  ```
+  OR
+  
+  If you need to connect to a remote k8s edge cluster, you can bind kcp to the public interface of your virtual machine:
+
+  ```shell
+  kcp start --bind-address <ip-address> >& kcp_log.txt &
+  ```
+  
+  This uses default 6443 port, just bind to a specific address
````
This is confusing as written.  "If you need to connect to a remote k8s edge cluster": why would the server-side address depend on whether the edge cluster is remote?  BTW, the kcp server does not open connections to the edge cluster.  I suspect the issue is about the syncer in the edge cluster being able to open connections to the kcp server.  But I do not understand why specifying a particular server address helps --- the default is to bind to 0.0.0.0, which is the least restrictive thing you can do.

"the public interface of your virtual machine" --- what virtual machine?  No virtual machine has been mentioned, at least not yet.

"This uses default 6443 port, just bind to a specific address" --- the port is potentially important regardless of whether a specific server address is bound.

MikeSpreitzer replied:
Perhaps the issue is the IP address that kcp writes into the generated kubeconfig.  That is never 0.0.0.0, it picks something more specific.

MikeSpreitzer commented on environments/dev-env/CONTRIBUTING.md:
```diff
@@ -0,0 +1,510 @@
+
+### 5. Connect your edge cluster to the kcp-edge platform:
+
+  a) Create inventory management workspace (`imw`):
```
"Create _an_ inventory management workspace"

dumb0002 replied:
fixed

MikeSpreitzer commented on environments/dev-env/CONTRIBUTING.md:
````diff
@@ -0,0 +1,510 @@
+  a) Create inventory management workspace (`imw`):
+
+  ```console
+  kubectl ws root
````
In a "console" block, we show the shell's prompts; this distinguishes input from output.

```console
$ kubectl ws root
Current context is "root"

$ kubectl ws create imw-1 --enter
Current context is "root:imw-1"
```

clubanderson replied:
my only reservation on that suggestion - adding a '$' renders the copy button useless - does it not?

MikeSpreitzer replied:
including the output, which we are not debating, makes the copy button useless

MikeSpreitzer replied:
We do not own the definition of the "console" tag.

dumb0002 replied:
Added. I looked at examples in other open source repos and they add the shell's prompts to  distinguishes input from output indeed.

MikeSpreitzer commented on environments/dev-env/CONTRIBUTING.md:
````diff
@@ -0,0 +1,510 @@
+  location.scheduling.kcp.io/florin labeled
+  ```
+
+  A location and synctarget objects are created:
````
I find the "are created" a bit vague.  I think it would be clearer to say that the preceding commands did this and the following commands show the effects.  Something like the following.

> The following commands list the objects that were created.

MikeSpreitzer commented on environments/dev-env/CONTRIBUTING.md:
````diff
@@ -0,0 +1,510 @@
+  ```
+
+
+  The [mailbox-controller](https://docs.kcp-edge.io/docs/coding-milestones/poc2023q1/mailbox-controller/) creates a mailbox workspace for the newly created SyncTarget: `florin`:
````
I think setting "`florin`" off with a colon says that `florin` is the mailbox workspace for the newly created SyncTarget.  I think it would be clearer to put `florin` in parentheses right after "SyncTarget".

dumb0002 replied:
fixed

MikeSpreitzer commented on environments/dev-env/CONTRIBUTING.md:
```diff
@@ -0,0 +1,510 @@
+  kubectl ws root
+  Current workspace is "root".
+
+  kubectl ws tree
+  .
+  └── root
+      ├── compute
+      ├── espw
+      │   └── 19igldm1mmolruzr-mb-6b0309f0-84f3-4926-9344-81df2f989f69
+      └── imw-1
```
This listing is not useful in general, when there is more than one mailbox workspace.  I think it would be better to do a listing like the one shown at the end of https://docs.kcp-edge.io/docs/coding-milestones/poc2023q1/example1/#the-mailbox-controller :

```console
$ kubectl ws root:espw
$ kubectl get Workspace -o "custom-columns=NAME:.metadata.name,SYNCTARGET:.metadata.annotations['edge\.kcp\.io/sync-target-name'],CLUSTER:.spec.cluster"
NAME                                                       SYNCTARGET   CLUSTER
apmziqj9p9fqlflm-mb-b8c64c64-070c-435b-b3bd-9c0f0c040a54   florin       12299slctppnhjnn
apmziqj9p9fqlflm-mb-bf452e1f-45a0-4d5d-b35c-ef1ece2879ba   guilder      yk9a66vjms1pi8hu
```

dumb0002 replied:
In this example for simplicity we are only deploying a single workload (`florin`). Then, we have only showing a single workspace. I like the tree structure because it clearly shows the workspace hierarchy used by kcp-edge

MikeSpreitzer replied:
We could show both.

francostellari commented:
I think a lot of comments on the MD are open to personal preferences in some cases and to different level of people reading.

In my opinion, I would try to get the script from Braulio's to kcp-edge main ASAP so that we can start referencing them correctly. They may not be perfect but they seem to work, at least in most cases, we can improve on them later.

To be honest I think that the script should go into `hack` not enviroment... and eventualy into the tar.gz

For the documentation, I think we need two nice an clear step by step pages into the kcp-edge.io more than in the git enviroment... This is not a dev-env anymore but the new quickstart for a) users that want to deploy b) users that want to develop and play with the details

MikeSpreitzer commented on environments/dev-env/CONTRIBUTING.md:
````diff
@@ -0,0 +1,510 @@
+
+  b) Start kcp:
+
+  ```shell
+  kcp start >& kcp_log.txt &
+  ```
+  OR
+  
+  If you need to connect to a remote k8s edge cluster, you can bind kcp to the public interface of your virtual machine:
+
+  ```shell
+  kcp start --bind-address <ip-address> >& kcp_log.txt &
+  ```
+  
+  This uses default 6443 port, just bind to a specific address
````
I suggest the following revision, for section (b) regarding starting kcp.

b) Start kcp:

```shell
kcp start --bind-address <ip-address> >& kcp_log.txt &
```

In that command, `<ip-address>` is an IP address of the machine running kcp that can be reached from the edge clusters.  This is the address that will appear in the generated kubeconfig file (mentioned next) and the kcp server will listen only at that address.  If all of the host's non-loopback addresses can be reached from the edge clusters then this is a non-issue and the `--bind-address <ip-address>` can be omitted from the command line; in this case the server will listen at all addresses and pick a non-loopback one to put in the generated kubeconfig file.

francostellari replied:
Perhaps `kcp start [--bind-address <ip-address>] >& kcp_log.txt &`

MikeSpreitzer replied:
That works for me.

MikeSpreitzer commented on environments/dev-env/build-edge-syncer.sh:
```diff
@@ -0,0 +1,56 @@
+#!/usr/bin/env bash
```
(1) Naming this script "build-<something>" is confusing, because what it does is not usually called "building".

(2) We already have `bin/mailbox-prep.sh` that gets the same job done.  I do not think we want to two scripts.  If the existing script lacks needed features (such as a `--verbose` flag) then let's fix that rather than make a substantially duplicate script.

MikeSpreitzer commented on environments/dev-env/build-edge-syncer.sh:
```diff
@@ -0,0 +1,56 @@
+if [ $verbosity == 1 ]; then
+    kubectl ws root:espw:$workspace
+    #kubectl kcp workload edge-sync $syncer_name --syncer-image $img -o $syncTarget_name-syncer.yaml    
+    ./kcp/bin/kubectl-kcp workload edge-sync $syncer_name --syncer-image $img -o $syncTarget_name-syncer.yaml
```
This script does not do anything to make sure the current working directory is anything in particular, so is not justified in assuming a particular meaning for the `.` at the start of this command line.

dumb0002 replied:
This script was created to be a helper functions of another main script only.

MikeSpreitzer commented:
I think we are agreed that there should be files in github that get used in the introductory examples.  This raises a basic structural problem: currently we have envisaged the first user experience to consist of downloading and installing releases of kcp and edge-mc, both containing only `bin` content, and then using those.  What about those example files?

It would be great if the example files could be left on github and just referenced by URL.  IIRC `kubectl` can read from a URL as well as from a file.  However, I suspect that not all of our example files are `kubectl` input.  For one example, if you look at example1.md you will see kind cluster config files (that write-up uses a cluster config file so that it can expose ports so that the deployed apps can be exercised by `curl` or a web browser on the developer's machine).  I suspect there are other cases.

To get the example files local in the user experience, I see a few plausible options.  One is that we abandon telling (either in detail or in a bash script) users to fetch and unpack a release archive, instead tell users as well as contributors to `git clone` the edge-mc repo.  But the problem with that is that we do not want to require users to build what they just `git clone`d.  But `git clone` does not yield built binaries.  One way around that would be to maintain the example files in a distinct repo.  The user experience would continue to include fetching and unpacking archives of binaries, and be extended with `git clone` of the edge-mc-examples repo.  Alternatively, we could also pack the example files into archives.  That could be either together with the binaries in one unified archive, or in distinct archives.  The former would be a bit weird, including examples in a release, but not totally unacceptable (IMO); it would have the advantage of clearly keeping consistency.  Using distinct archives makes more steps for the user, although in the quickstart this can be under the covers of a big wrapper script.

dumb0002 commented:
@MikeSpreitzer, yes indeed. Let's discuss and propose how we should merge the content of both CONTRIBUTING.md files

dumb0002 commented:
Some samples repo for developer experience:

Etcd: 
   - https://github.com/etcd-io/etcd/blob/main/CONTRIBUTING.md

kyverno:
  - https://github.com/kyverno/kyverno/blob/main/DEVELOPMENT.md

Should we maybe call our file DEVELOPMENT.md? Most of the CONTRIBUTING.md file that I found on public repos were focused on providing the guidelines to write code and create pull requests, etc.

dumb0002 commented on environments/dev-env/CONTRIBUTING.md:
````diff
@@ -0,0 +1,510 @@
+./install_req.sh
+```
+
+Alternatively, run the following command for MacOS platform:
````
fixed

dumb0002 commented:
@MikeSpreitzer 


> Shouldn't this PR be updating the existing CONTRIBUTING.md rather than creating an alternate file?


I created a new file DEVELOPMENT.md in the main directory of this repo, as I have seen in other open source repos, for example: https://github.com/knative/serving

dumb0002 commented:
> I think we are agreed that there should be files in github that get used in the introductory examples. This raises a basic structural problem: currently we have envisaged the first user experience to consist of downloading and installing releases of kcp and edge-mc, both containing only `bin` content, and then using those. What about those example files?
> 
> It would be great if the example files could be left on github and just referenced by URL. IIRC `kubectl` can read from a URL as well as from a file. However, I suspect that not all of our example files are `kubectl` input. For one example, if you look at example1.md you will see kind cluster config files (that write-up uses a cluster config file so that it can expose ports so that the deployed apps can be exercised by `curl` or a web browser on the developer's machine). I suspect there are other cases.
> 
> To get the example files local in the user experience, I see a few plausible options. One is that we abandon telling (either in detail or in a bash script) users to fetch and unpack a release archive, instead tell users as well as contributors to `git clone` the edge-mc repo. But the problem with that is that we do not want to require users to build what they just `git clone`d. But `git clone` does not yield built binaries. One way around that would be to maintain the example files in a distinct repo. The user experience would continue to include fetching and unpacking archives of binaries, and be extended with `git clone` of the edge-mc-examples repo. Alternatively, we could also pack the example files into archives. That could be either together with the binaries in one unified archive, or in distinct archives. The former would be a bit weird, including examples in a release, but not totally unacceptable (IMO); it would have the advantage of clearly keeping consistency. Using distinct archives makes more steps for the user, although in the quickstart this can be under the covers of a big wrapper script.


- I removed the files related to example1.md  in this PR. I may need to ask the users to clone the repo in order to use the stage scenarios in our automation using the `kcp-playground` tool (aka demo environment)

dumb0002 commented:
FYI, I will create another PR for the following scripts cited in this instruction: `kcp-edge.sh`, `install_req.sh`, `clean_up.sh`

MikeSpreitzer commented:
As another example of documentation for developers, here is what the Kubernetes project has: https://github.com/kubernetes/community/blob/master/contributors/devel/README.md .  It is much more elaborate than we should have at this point.

MikeSpreitzer commented on DEVELOPMENT.md:
```diff
@@ -0,0 +1,496 @@
+# KCP-Edge Contributor Development Environment
```
I think the title needs adjusting; we are no longer talking about a distinct "environment" for developers, right?

dumb0002 replied:
yes, we'll still have separated instructions for users versus developers, aren't we?  please, clarify. thanks!

MikeSpreitzer replied:
How about simply "Contributor Instructions"?

MikeSpreitzer replied:
"Users" may do 'development' too --- development of higher level code.  That is why I use the terms "contributor" for developers of this package and "users".

dumb0002 replied:
Fixed

MikeSpreitzer commented on DEVELOPMENT.md:
```diff
@@ -0,0 +1,496 @@
+  - Windows WSL/Ubuntu
+
+
+After cloning this repo (see quickstart), run the following script to install the required package (Linux or MacOS):
```
I suppose the intended referent is the "### 1. Clone the kcp-edge repo" section below; the existing reference to "quickstart" does not obviously lead there.

MikeSpreitzer commented on DEVELOPMENT.md:
````diff
@@ -0,0 +1,496 @@
+kind create cluster --name florin
+```  
+
+### 4. Install KCP
````
I think that at this point the reader should be referred to the user documentation.
Attempting to maintain a copy here adds work for us and opportunities to make mistakes.
Having a copy of a large volume of material here also confuses the developer: she wonders why she is reading something different from users --- is there a reason?  Is there an important difference that she is overlooking?

MikeSpreitzer commented on DEVELOPMENT.md:
````diff
@@ -0,0 +1,496 @@
+```
+
+
+### 3. Create your own edge infrastructure (edge clusters) - Bring Your Own Clusters (BYOC)
````
I think that at this point the reader should be referred to the user documentation.
Attempting to maintain a copy here adds work for us and opportunities to make mistakes.
Having a copy of a large volume of material here also confuses the developer: she wonders why she is reading something different from users --- is there a reason?  Is there an important difference that she is overlooking?

dumb0002 replied:
What steps would be included in the developer page? just the steps to clone a repo and run `make build`?

I was thinking that the developer documentation would be self-contained in a single page.

dumb0002 replied:
Maybe we don't need the same level of detail in the user instructions. For example, we don't expose in the user's instructions the concepts of `mailbox workspace`, `SinglePlacementSlice` and `syncerConfig`.

dumb0002 commented:
@MikeSpreitzer

> When it comes to "cleanup": the reader wonders what exactly is being removed. Please at least say that clearly. There is a big difference between removing an example usage and uninstalling the platform. Perhaps give the scripts more specific names?
> 

I updated the instructions to clearly explain what's being cleaned or removed. 

> Also, think about the granularity desired --- making tear-down less granular than set-up means potentially un-doing and re-doing work. I had the unhappy experience of doing a lot of that last year.

I added more explanation and a new instruction describing the granularity of the tear-down

dumb0002 commented:
@MikeSpreitzer,  here is the PR for scripts described in these instructions: https://github.com/kcp-dev/edge-mc/pull/354

MikeSpreitzer commented on DEVELOPMENT.md:
```diff
@@ -0,0 +1,94 @@
+  - Windows WSL/Ubuntu
+
+
+After cloning this repo (see [instructions](#1-clone-the-kcp-edge-repo)), run the following script to install the required package (Linux or MacOS):
```
I think "package" should be plural here.

MikeSpreitzer commented on DEVELOPMENT.md:
````diff
@@ -0,0 +1,94 @@
+After cloning this repo (see [instructions](#1-clone-the-kcp-edge-repo)), run the following script to install the required package (Linux or MacOS):
+
+```bash
+hack/install_req.sh
````
(1) I do not see this file.
(2) Does it install Docker?  I suspect not.  If I am right then the user should not be told that this script will simply "install the required packages".

MikeSpreitzer commented on DEVELOPMENT.md:
````diff
@@ -0,0 +1,94 @@
+On MacOS you could use the following command as an alternative to the above:
+
+```bash
+brew install ko gcc jq make go@1.19 kind kubectl
````
I see no Docker here.

MikeSpreitzer commented on DEVELOPMENT.md:
````diff
@@ -0,0 +1,94 @@
+### 2. Build the kcp-edge binaries:
+
+```shell
+cd edge-mc
````
I would move the `cd` into step 1; it is a bit of context-setting that should _not_ be done as the contributor iterates through cycles of edit-build-test.

MikeSpreitzer commented on DEVELOPMENT.md:
````diff
@@ -0,0 +1,94 @@
+make build
+```
+
+The kcp-edge binaries is created inside the `./bin` directory. Add this directory to your $PATH:
````
Not everything created by `make build` is a "binary".  Remember that scripts also go into `bin`.  That is why I say "executable" instead.

MikeSpreitzer commented on DEVELOPMENT.md:
````diff
@@ -0,0 +1,94 @@
+
+  a) Download kcp binaries for your platform:
+
+  ```shell
+  hack/install-kcp-with-plugins.sh --version v0.11.0 --folder $(pwd)/kcp --create-folder
+  ```
````
I would point out that this script is just an automation of the instructions at https://docs.kcp.io/kcp/v0.11/#download-kcp so that the reader knows that KubeStellar does not require anything special and gets good description of what the script does without having to fetch and read the script.

MikeSpreitzer commented on DEVELOPMENT.md:
````diff
@@ -0,0 +1,94 @@
+  ```
+  OR
+  
+  If you need to connect to a remote k8s edge cluster, you can bind kcp to the public interface of your virtual machine:
````
Sometimes when I am reading instructions in a hurry, I do things in the order I am told.  If the instructions then say "but wait, you should have done something different if FOO" then I am already in trouble --- I have done the wrong thing before I read that amendment.  I recommend rewording this so as to not lead with something that we quickly retract.

MikeSpreitzer commented:
The curl-to-bash technique that the quickstart says to use is convenient but wholly inadequate both for contributors and for users who are building KubeStellar into their own bigger system.  We should have documentation of how to install and launch kcp and KubeStellar, and I think that we should have one document that is usable by both of the categories of people listed above (because most of the content is the same).  The only difference between them would be between downloading and unpacking a KubeStellar release distribution vs git clone and build.  

What I see in DEVELOPMENT.md now is almost exactly what I just asked for.  Maybe we should make it that?
